Start with a small example that moves functions.

alex4o opened the pull request:
It moves functions from one bitcode file to another. But it still
doesn't work as the resulting file cannot be compiled with llc.

ltratt commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+
```
Extraneous blank line.

ltratt replied:
I think this one hasn't been fixed?

ltratt commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+static llvm::LLVMContext context;
+
+int main() {
+	std::cout << "Hello world" << std::endl;
```
Probably not necessary ;)

ltratt commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+int main() {
+	std::cout << "Hello world" << std::endl;
+	llvm::SMDiagnostic err;
+	auto file =  llvm::StringRef("/home/alex4o/bitcode_experiments_cxx/file.bc");
```
Hard-coded path.

ltratt commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+				llvm::CloneFunctionInto(resultFunction, &function, VMap, false, returns);
+
+				for(llvm::ReturnInst* inst : returns) {
+					//inst->print(llvm::outs(), false);
```
Commented code? Maybe the loop can be removed?

alex4o commented:
I am leaving this here, because until now, I sued shared but it might be better to use internal. 
https://lists.llvm.org/pipermail/cfe-dev/2015-October/045580.html

alex4o commented:
Also ready for review.

ltratt commented:
> I am leaving this here, because until now, I sued shared but it might be better to use internal.

Not sure I quite understand.

alex4o commented:
So until now I have used `__attribute__ ((visibility ("hidden")))` but I found this thread that argues that `__attribute__ ((internal_linkage))` is a more precise choice. Both of these things are for `C` but they have their respective representations in the llvm IR. I was hoping you would have an opinion on that, but it is not mandatory.

ltratt commented:
> this thread

No link? And I'm not sure if this related to "sued shared" or not :)

alex4o commented:
Ah sorry, I reworded the comment.

ltratt commented on main.cpp:
```diff
@@ -24,46 +26,54 @@
+        llvm::Function *resultFunction =
+            llvm::dyn_cast_or_null<llvm::Function>(resultCallee.getCallee());
+        if(resultFunction == nullptr) {
+          std::cout << "ERROR" << std::endl;
```
Should we be failing on the first error rather than burying it?

alex4o replied:
That code got removed in the later commits, as it was rewritten in a way that couldn't fail.

ltratt commented on main.cpp:
```diff
@@ -24,46 +26,54 @@
-				}
-				
-				llvm::SmallVector<llvm::ReturnInst*, 8> returns;
+  llvm::ValueToValueMapTy VMap;
```
This may by lack of understanding of C++ standards, but `VMap` thus capitalised was a slightly confusing variable name as I thought it was a type below!

alex4o replied:
This is how the code was in the original CloneModule source code. But I understand what you mean, I would probably make all the variables start with a lowercase.

ltratt replied:
Ah, have we copy and pasted this from somewhere?

alex4o replied:
I am more or less trying to replicate this function: https://llvm.org/doxygen/CloneFunction_8cpp_source.html but doing it with our requirements in mind.

ltratt commented on file.c:
```diff
@@ -0,0 +1,21 @@
+//#include<stdio.h>
```
Commented code.

ltratt commented on file.c:
```diff
@@ -0,0 +1,21 @@
+}
+
+int main() {
+//	printf("%d %d\n", b(), c());
```
Commented code.

ltratt commented on main.cpp:
```diff
@@ -49,24 +50,42 @@ void moveFunctions(const llvm::StringRef outputName, const llvm::Module *module,
 
   llvm::ValueToValueMapTy VMap;
 
+  for (const auto &alias : module->aliases()) {
```
I'm not quite sure what this loop is doing. Maybe add a brief comment?

ltratt commented on main.cpp:
```diff
@@ -49,24 +50,42 @@ void moveFunctions(const llvm::StringRef outputName, const llvm::Module *module,
+        nullptr, global.getName(), nullptr, global.getThreadLocalMode(),
         global.getType()->getAddressSpace());
 
+    //GV->setAlignment(global.getAlign());
```
Commented code.

ltratt commented on main.cpp:
```diff
@@ -88,6 +104,18 @@ void moveFunctions(const llvm::StringRef outputName, const llvm::Module *module,
+      continue;
+    }
+
+    // std::cout << "Cloning: " << action << " " << function.getName().str()          << std::endl;
```
Commented code.

ltratt commented on main.cpp:
```diff
@@ -100,8 +128,18 @@ void moveFunctions(const llvm::StringRef outputName, const llvm::Module *module,
+    copyComdat(resultFunction, &function);
   }
 
+     // And aliases
```
Too much indentation?

ltratt commented on main.cpp:
```diff
@@ -118,8 +156,12 @@ int main() {
+
   if (m) {
-    moveFunctions(
+    /*moveFunctions(
```
Commented code?

ltratt commented:
I've left a few (mostly very shallow!) comments. One thing I'd like to encourage is a few comments (they don't have to be long) in the code giving a few hints as to what e.g. functions are doing. I had to do a fair bit of guessing and I'm guessing I guessed wrong several times :)

alex4o commented:
Let me ask a question, when you have left comments on files that were already changed, should I ignore them, or fix those commits?

ltratt commented:
I think commented code should never be in the commits, so you should probably edit history for those so that we appear professional to the future :) But, for "semantic" changes that are fixed by later commits, you shouldn't rewrite history. Sound like a plan?

cristianurlea commented:
> So until now I have used `__attribute__ ((visibility ("hidden")))` but I found this thread that argues that `__attribute__ ((internal_linkage))` is a more precise choice. Both of these things are for `C` but they have their respective representations in the llvm IR. I was hoping you would have an opinion on that, but it is not mandatory.

For the C example given ( file.c ) i *believe* it is sufficient to make `int a()` static. This will cause it to be marked for internal linkage. The use of the `hidden` and `internal_linkage` attributes is only required for C++ methods because the `static` keyword has a different meaning in that context.

cristianurlea commented on main.cpp:
```diff
@@ -0,0 +1,205 @@
+        nullptr, global.getName(), nullptr, global.getThreadLocalMode(),
+        global.getType()->getAddressSpace());
+
+    // TODO: Find a way to understand how CloneModule copies these things
```
I suspect that `CloneModule` is meant for use in the earlier stages of compilation, prior to the target-specific passes which typically set the required alignment, and so does attempt to preserve alignment.

alex4o commented:
I rewrote the history, should I force push?

ltratt commented:
Yes please!

alex4o commented:
I did but the dates are all wrong now, but this should be fine now I hope.

alex4o commented:
I think this should be fine for review, I hope.

ltratt commented:
Please squash.

alex4o commented:
As I am going to squash everything into a single commit I have decided to go through the git log and see which messages are worth keeping as comments, I also deleted some of the code that looks like it doesn't contribute to the final result, but I am afraid that not understanding what I removed might be a problem later. But everything works fine now. As I have added additional commits should they be reviewed as well?

ltratt commented:
I've reviewed the new commit -- we're good!

I think squashing to a single commit is the right thing in this case. If you use `s` in `git rebase` it'll give you all the old commit messages and then you can pick and choose the bits that make most sense.

alex4o commented:
We can merge now.

ltratt commented:
Have we set up buildbot for this repo?

alex4o commented:
Not at the moment, but as I am going to work longer on it, this is something I will do soon.

ltratt commented:
Shall we do it for this PR or after?

alex4o commented:
I would require, modifying the buildbot config I thing so let's do it in a second pr.

ltratt commented:
OK.
